Add HoC replacement of update-post-status mixin; use in Pages

mcsf opened the pull request:
~Branch based on [`fix/pages-popover-highlight`](https://github.com/Automattic/wp-calypso/pull/15152). [See direct diff](https://github.com/Automattic/wp-calypso/compare/fix/pages-popover-highlight...remove/update-post-status-mixin?expand=1).~ [merged]

# Why

- We'd like Calypso to be free of mixins soon-ish, and a few of these can be replaced with analogous higher-order components (HoC).
- Pages is crumbling under tech debt; mixin usage is just a sliver, but entangling it from `mixins/update-post-status` is a step in the right direction.

# Reviewing

Right now, CI is *not* happy about the non-namespaced usage of `className`. This PR is large as it is; I'm not sure that that work, which would include making sure no CSS rule is affected throughout our assets, should be covered by it.

# Testing

In `/pages`, across its different tabs (published, draft, scheduled, trash), make sure that page actions (via the ellipsis button) work the same way as before. Example:
- Trashing a page successfully
- Undoing that action
- Trashing it again, without undoing
- Navigating to Trash
- Restoring it
- Trashing again + actually deleting
- etc.

mcsf commented:
@ockham & @gwwar: I assigned you as reviewers tentatively in order to get some shared knowledge (as well as great reviews, as usual!) on this, but feel free to dismiss. 🙇

gwwar commented:
@mcsf I'll try to take a 👀  tomorrow

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+import analytics from 'lib/analytics';
+
+// TODO(mcsf): use i18n#localize?
+const strings = {
```
While we do a hard refresh when a user changes language settings, I think this is safer if this was translated in a lifecycle event, or

mcsf replied:
Sounds good.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+				status={ this.state.updatedStatus }
+				strings={ strings[ this.getType() ] }
+			/>;
+		}
```
Complete aside, but why do folks prefer mixed usage of `;` for class properties? eg why `state = {};` but not for `foo = () => {};` ? Feel free to ignore this comment though, this is perhaps best left for another styleguide PR.

mcsf replied:
To be honest I wouldn't mind ditching most semi-colons in JS in general. In this case, I might remove them from `state` and `displayName`.

gwwar replied:
Either way works for me tbh. The babel transform example does it that way, since the semis are added afterwards. I'd just like us to be consistent here.

gwwar commented on client/my-sites/pages/page/index.js:
```diff
@@ -50,10 +48,17 @@ function recordEvent( eventAction ) {
 
 // FIXME(mcsf): I vow to follow-up on this and port to Component
 const Page = React.createClass( { // eslint-disable-line react/prefer-es6-class
```
Perhaps remove the disable line, since we do want to clean this up later and the tests are already red.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+				? { group: 'Pages', eventName: 'Clicked Undo Trashed Page' }
+				: { group: 'Posts', eventName: 'Clicked Undo Trashed Post' };
+
+			analytics.ga.recordEvent( group, eventName );
```
Let's avoid direct analytics calls.

gwwar commented on client/my-sites/pages/page/index.js:
```diff
@@ -438,34 +420,30 @@ const Page = React.createClass( { // eslint-disable-line react/prefer-es6-class
-		this.setState( { showPageActions: false } );
-		this.updatePostStatus( 'delete' );
+		this.props.updatePostStatus( 'delete' );
 		recordEvent( 'Clicked Delete Page' );
```
If it's not too much trouble can we swap this for
```jsx
const mapDispatchToProps: {
   recordEvent: partial( recordGoogleEvent, 'Pages' ),
//...
}
```

Though I'm fine with deferring this for later if we're trying to keep the diff size down.

mcsf replied:
Good point, but will leave for another PR.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+			const post = this.props.post || this.props.page;
+			let previousStatus = null;
+
+			const setNewStatus = ( error, resultPost ) => {
```
Looks like we only need the closure here for `previousStatus`. This may be worth refactoring later.

mcsf replied:
Indeed, a lot could be altogether thought anew, but the focus was on getting a working HoC based on the mixin. Maybe some of this could even just rely on basic action dispatching + state selection.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+				} );
+
+				const type = this.props.post ? 'post' : 'page';
+				if ( window.confirm( strings[ type ].deleteWarning ) ) { // eslint-disable-line no-alert
```
Let's check for window before using it. This is helpful if we ever want to test this in the future.

mcsf replied:
Agreed.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+				return true;
+			};
+
+			if ( status === 'delete' ) {
```
A switch might be a better fit for these status checks, but completely optional.

gwwar commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,188 @@
+				if ( ! setNewStatus( error, resultPost ) ) {
+					return;
+				}
+				setTimeout( this.resetState, 1200 );
```
Perhaps `const RESET_TIMEOUT_MS = 1200;` ?

gwwar commented on client/components/update-post-status/update-template.js:
```diff
@@ -0,0 +1,72 @@
+			undoTemplate = (
+				<a className="undo" onClick={ undoClick }><span>{ s.undo }</span></a>
+			);
+			updateText = s.trashed;
```
I guess we were trying to be clever here for page vs post. I suppose we should know the type from the full post object?

So we can maybe update this to
```jsx
// trashed
updateText = translate( 'Moved to Trash' );

// deleteWarning
updateText = translate( 'Delete this %(postType)s permanently?', { args: { postType } } );
```

gwwar replied:
Hmm, I suppose we'd need a translated postType for the above example. It might be difficult without context.

mcsf replied:
Yeah. Post types and i18n are hard — or easy but very inflexible. There's a lot of reasons why that wouldn't go well. Example off the top of my head:

> Delete this %(postType)s permanently?

→ Apagar **este** artigo para sempre?
→ Apagar **esta** página para sempre?

This would be `pt-pt`, which features gendered nouns, like many other Romance languages, but other features can break this as well. I think the current impl is a _good enough_ solution that I'd like to leave untouched for now.

mcsf commented:
@gwwar: thanks for the review! I've addressed all your feedback except when noted otherwise. If you have some spare time, feel free to have a last look at the diff after the changes, otherwise I'll merge this sometime around 1600UTC.

mcsf commented on client/components/update-post-status/index.js:
```diff
@@ -0,0 +1,207 @@
+				return;
+			}
+
+			const strings = getStrings( this.props.translate );
```
Not the most elegant, but I'm convinced it will do for now. Trying not to waste too much time on the HoC and focus on getting rid of the mixin.
